File: kms-samples-python/envelope_decrypt.py
import argparse
import base64
import json

from Crypto.Cipher import AES
from aliyunsdkcore.client import AcsClient
from aliyunsdkkms.request.v20160120 import DecryptRequest


def kms_decrypt(client, cipher_text):
    request = DecryptRequest.DecryptRequest()
    request.set_accept_format('JSON')
    request.set_CiphertextBlob(cipher_text)
    response = json.loads(client.do_action_with_exception(request))
    return response.get('Plaintext')


def read_text_file(in_file):
    with open(in_file, 'r') as f:
        lines = []
        for line in f:
            lines.append(line)
    return lines


def write_text_file(out_file, content):
    with open(out_file, 'w') as f:
        f.write(content)


def local_decrypt(data_key, iv, cipher_text, tag, out_file):
    cipher = AES.new(data_key, AES.MODE_GCM, iv)
    data = cipher.decrypt_and_verify(cipher_text, tag)
    write_text_file(out_file, data.decode('utf-8'))


def main():
    parser = argparse.ArgumentParser()
    parser.add_argument('--ak', help='the access key id')
    parser.add_argument('--as', help='the access key secret')
    parser.add_argument('--region', default='cn-hangzhou', help='the region id')
    args = vars(parser.parse_args())

    client = AcsClient(args["ak"], args["as"], args["region"])
    # client.set_verify(False)

    in_file = './data/sales.csv.cipher'
    out_file = './data/decrypted_sales.csv'

    # Read encrypted file
    in_lines = read_text_file(in_file)

    # Decrypt data key
    data_key = kms_decrypt(client, in_lines[0])

    # Locally decrypt the sales record
    local_decrypt(base64.b64decode(data_key),
                  base64.b64decode(in_lines[1]),
                  base64.b64decode(in_lines[2]),
                  base64.b64decode(in_lines[3]),
                  out_file)


if __name__ == '__main__':
    main()

File: kms-samples-java/src/main/java/com/aliyun/kms/samples/EnvelopeDecrypt.java
package com.aliyun.kms.samples;

import com.aliyuncs.DefaultAcsClient;
import com.aliyuncs.exceptions.ClientException;
import com.aliyuncs.http.FormatType;
import com.aliyuncs.http.HttpClientConfig;
import com.aliyuncs.http.MethodType;
import com.aliyuncs.http.ProtocolType;
import com.aliyuncs.kms.model.v20160120.DecryptRequest;
import com.aliyuncs.kms.model.v20160120.DecryptResponse;
import com.aliyuncs.profile.DefaultProfile;
import com.aliyuncs.profile.IClientProfile;

import javax.crypto.Cipher;
import javax.crypto.spec.GCMParameterSpec;
import javax.crypto.spec.SecretKeySpec;
import java.io.*;
import java.nio.file.Files;
import java.nio.file.Paths;
import java.util.Base64;
import java.util.List;

public class EnvelopeDecrypt {
    private static DefaultAcsClient kmsClient;
    private static final int GCM_TAG_LENGTH = 16;

    private static DefaultAcsClient kmsClient(String regionId, String accessKeyId, String accessKeySecret) {
        IClientProfile profile = DefaultProfile.getProfile(regionId, accessKeyId, accessKeySecret);
        HttpClientConfig clientConfig = HttpClientConfig.getDefault();
        //clientConfig.setIgnoreSSLCerts(true);
        profile.setHttpClientConfig(clientConfig);

        return new DefaultAcsClient(profile);
    }

    private static String kmsDecrypt(String cipherTextBlob) throws ClientException {
        final DecryptRequest request = new DecryptRequest();
        request.setSysProtocol(ProtocolType.HTTPS);
        request.setAcceptFormat(FormatType.JSON);
        request.setSysMethod(MethodType.POST);
        request.setCiphertextBlob(cipherTextBlob);
        DecryptResponse response = kmsClient.getAcsResponse(request);
        return response.getPlaintext();
    }

    private static List<String> readTextFile(String inFile) throws IOException {
        return Files.readAllLines(Paths.get(inFile));
    }

    private static void writeTextFile(String outFile, byte[] content) throws IOException {
        File file = new File(outFile);
        try (FileOutputStream fos = new FileOutputStream(file)) {
            fos.write(content);
        }
    }

    private static void localDecrypt(byte[] dataKey, byte[] iv, byte[] cipherText, String outFile) throws Exception {
        Cipher cipher = Cipher.getInstance("AES/GCM/NoPadding");
        SecretKeySpec keySpec = new SecretKeySpec(dataKey, "AES");
        GCMParameterSpec gcmParameterSpec = new GCMParameterSpec(GCM_TAG_LENGTH * 8, iv);
        cipher.init(Cipher.DECRYPT_MODE, keySpec, gcmParameterSpec);
        byte[] decryptedText = cipher.doFinal(cipherText);
        writeTextFile(outFile, decryptedText);
    }

    public static void main(String[] args) {
        String regionId = "cn-hangzhou";
        String accessKeyId = System.getenv("AccessKeyId");
        String accessKeySecret = System.getenv("AccessKeySecret");

        kmsClient = kmsClient(regionId, accessKeyId, accessKeySecret);

        String inFile = "./data/sales.csv.cipher";
        String outFile = "./data/decrypted_sales.csv";

        try{
            //Read encrypted file
            List<String> inLines = readTextFile(inFile);

            //Decrypt data key
            String dataKey = kmsDecrypt(inLines.get(0));

            //Locally decrypt the sales record
            localDecrypt(Base64.getDecoder().decode(dataKey),
                    Base64.getDecoder().decode(inLines.get(1)),
                    Base64.getDecoder().decode(inLines.get(2)),
                    outFile);

        } catch (ClientException e) {
            System.out.println("Failed.");
            System.out.println("Error code: " + e.getErrCode());
            System.out.println("Error message: " + e.getErrMsg());
        } catch (Exception e) {
            e.printStackTrace();
        }
    }
}
